File: csharp/src/json/JsonBuilder.cs
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

using System.Diagnostics;
using System.Text;

namespace finalmq
{
    public class JsonBuilder : IJsonParserVisitor
    {
        static byte[] m_strNull = { (byte)'n', (byte)'u', (byte)'l', (byte)'l', (byte)',' };
        static byte[] m_strTrue = { (byte)'t', (byte)'r', (byte)'u', (byte)'e', (byte)',' };
        static byte[] m_strFalse = { (byte)'f', (byte)'a', (byte)'l', (byte)'s', (byte)'e', (byte)',' };
        public JsonBuilder(IZeroCopyBuffer buffer, int maxBlockSize = 512)
        {
            m_zeroCopybuffer = buffer;
            m_maxBlockSize = maxBlockSize;
        }

        // IJsonParserVisitor
        public void SyntaxError(string buffer, string message)
        {
            Debug.Assert(false);
        }
        public void EnterNull()
        {
            ReserveSpace(5);
            Debug.Assert(m_buffer != null);
            m_strNull.CopyTo(m_buffer, m_offset);
            m_offset += 5;
        }
        public void EnterBool(bool value)
        {
            ReserveSpace(6);
            Debug.Assert(m_buffer != null);
            if (value)
            {
                m_strTrue.CopyTo(m_buffer, m_offset);
                m_offset += 5;
            }
            else
            {
                m_strFalse.CopyTo(m_buffer, m_offset);
                m_offset += 6;
            }
        }
        public void EnterInt32(int value)
        {
            ReserveSpace(20);   // 11 + 1 is enough: 10 digits + 1 minus + 1 comma
            Debug.Assert(m_buffer != null);
            string s = value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            m_offset += Encoding.UTF8.GetBytes(s, 0, s.Length, m_buffer, m_offset);
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterUInt32(uint value)
        {
            ReserveSpace(20);   // 10 + 1 is enough: 10 digits + 1 comma
            Debug.Assert(m_buffer != null);
            string s = value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            m_offset += Encoding.UTF8.GetBytes(s, 0, s.Length, m_buffer, m_offset);
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterInt64(long value)
        {
            ReserveSpace(30);   // 21 + 1 is enough: 20 digits + 1 minus + 1 comma
            Debug.Assert(m_buffer != null);
            string s = value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            m_offset += Encoding.UTF8.GetBytes(s, 0, s.Length, m_buffer, m_offset);
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterUInt64(ulong value)
        {
            ReserveSpace(30);   // 20 + 1 is enough: 20 digits + 1 comma
            Debug.Assert(m_buffer != null);
            string s = value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            m_offset += Encoding.UTF8.GetBytes(s, 0, s.Length, m_buffer, m_offset);
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterDouble(double value)
        {
            ReserveSpace(50);
            Debug.Assert(m_buffer != null);
            string s = value.ToString(System.Globalization.CultureInfo.InvariantCulture);
            m_offset += Encoding.UTF8.GetBytes(s, 0, s.Length, m_buffer, m_offset);
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterString(string value)
        {
            SerializeStringWithEnding(value, (byte)',');
        }
        public void EnterString(byte[] buffer, int offset, int size)
        {
            SerializeStringWithEnding(buffer, offset, size, (byte)',');
        }
        public void EnterArray()
        {
            ReserveSpace(1);
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)'[';
            ++m_offset;
        }
        public void ExitArray()
        {
            CorrectComma();
            ReserveSpace(2);
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)']';
            ++m_offset;
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterObject()
        {
            ReserveSpace(1);
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)'{';
            ++m_offset;
        }
        public void ExitObject()
        {
            CorrectComma();
            ReserveSpace(2);
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)'}';
            ++m_offset;
            m_buffer[m_offset] = (byte)',';
            ++m_offset;
        }
        public void EnterKey(string key)
        {
            SerializeStringWithEnding(key, (byte)':');
        }
        public void EnterKey(byte[] buffer, int offset, int size)
        {
            SerializeStringWithEnding(buffer, offset, size, (byte)':');
        }
        public void Finished()
        {
            CorrectComma();
            ResizeBuffer();
        }

        void ReserveSpace(int space)
        {
            int sizeRemaining = m_offsetEnd - m_offset;
            if (sizeRemaining < space)
            {
                if (m_bufferStart != null)
                {
                    int size = m_offset - m_bufferStart.Offset;
                    Debug.Assert(size >= 0);
                    m_zeroCopybuffer.DownsizeLastBuffer(size);
                }

                int sizeRemainingZeroCopyBuffer = m_zeroCopybuffer.RemainingSize;
                int sizeNew = m_maxBlockSize;
                if (space <= sizeRemainingZeroCopyBuffer)
                {
                    sizeNew = sizeRemainingZeroCopyBuffer;
                }
                sizeNew = Math.Max(sizeNew, space);
                BufferRef bufferStartNew = m_zeroCopybuffer.AddBuffer(sizeNew);
                m_buffer = bufferStartNew.Buffer;
                m_bufferStart = bufferStartNew;
                m_offsetEnd = m_bufferStart.Offset + sizeNew;
                m_offset = m_bufferStart.Offset;
            }
        }
        void ResizeBuffer()
        {
            if (m_bufferStart != null)
            {
                int size = m_offset - m_bufferStart.Offset;
                Debug.Assert(size >= 0);
                m_zeroCopybuffer.DownsizeLastBuffer(size);
                m_bufferStart = null;
                m_buffer = null;
                m_offsetEnd = 0;
                m_offset = 0;
            }
        }
        void CorrectComma()
        {
            if (m_bufferStart != null)
            {
                Debug.Assert(m_buffer != null);
                int size = m_offset - m_bufferStart.Offset;
                Debug.Assert(size >= 0);
                if (size >= 1)
                {
                    if (m_buffer[m_offset - 1] == ',')
                    {
                        --m_offset;
                    }
                }
            }
        }
        void EscapeString(string str)
        {
            Debug.Assert(m_buffer != null);

            for (int i = 0; i < str.Length; i++)
            {
                char c = str[i];
                if (!EscapeByte(c))
                {
                    m_buffer[m_offset] = (byte)'\\';
                    m_offset++;
                    m_buffer[m_offset] = (byte)'u';
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[((c >> 12) & 0x0f)];
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[((c >> 8) & 0x0f)];
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[((c >> 4) & 0x0f)];
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[(c & 0x0f)];
                    m_offset++;
                }
            }
        }

        bool EscapeByte(char c)
        {
            Debug.Assert(m_buffer != null);
            if (c < 0x20)
            {
                m_buffer[m_offset] = (byte)'\\';
                m_offset++;
                byte e = escape[c];
                m_buffer[m_offset] = e;
                m_offset++;
                if (e == 'u')
                {
                    m_buffer[m_offset] = (byte)'0';
                    m_offset++;
                    m_buffer[m_offset] = (byte)'0';
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[(c >> 4)];
                    m_offset++;
                    m_buffer[m_offset] = hexDigits[(c & 0x0f)];
                    m_offset++;
                }
                return true;
            }
            else if (c < 0x80)
            {
                if (c == '\\' || c == '\"')
                {
                    m_buffer[m_offset] = (byte)'\\';
                    m_offset++;
                }
                m_buffer[m_offset] = (byte)c;
                m_offset++;
                return true;
            }
            return false;
        }
        void SerializeStringWithEnding(string value, byte ending)
        {
            int size = value.Length;
            ReserveSpace(size * 6 + 2 + 1); // string*6 + 2" + ending (',' or ':')
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)'\"';
            ++m_offset;
            EscapeString(value);
            m_buffer[m_offset] = (byte)'\"';
            ++m_offset;
            m_buffer[m_offset] = ending;
            ++m_offset;
        }
        void SerializeStringWithEnding(byte[] buffer, int offset, int size, byte ending)
        {
            ReserveSpace(size * 6 + 2 + 1); // string*6 + 2" + ending (',' or ':')
            Debug.Assert(m_buffer != null);
            m_buffer[m_offset] = (byte)'\"';
            ++m_offset;
            bool escapeNeeded = false;
            int end = offset + size;
            int i = offset;
            for (; i < end; ++i)
            {
                byte c = buffer[i];
                if (!EscapeByte((char)c))
                {
                    escapeNeeded = true;
                    break;
                }
            }
            if (escapeNeeded)
            {
                EscapeString(Encoding.UTF8.GetString(buffer, i, end - i));
            }
            m_buffer[m_offset] = (byte)'\"';
            ++m_offset;
            m_buffer[m_offset] = ending;
            ++m_offset;
        }

        IZeroCopyBuffer m_zeroCopybuffer;
        int m_maxBlockSize = 512;
        BufferRef? m_bufferStart = null;
        byte[]? m_buffer;
        int m_offset = 0;
        int m_offsetEnd = 0;

        static byte[] hexDigits = new byte[] { (byte)'0', (byte)'1', (byte)'2', (byte)'3', (byte)'4', (byte)'5', (byte)'6', (byte)'7', (byte)'8', (byte)'9', (byte)'a', (byte)'b', (byte)'c', (byte)'d', (byte)'e', (byte)'f' };

        static byte[] escape = new byte[0x20]
        {
                //0        1          2          3          4          5          6          7          8          9          A          B          C          D          E          F
                (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'b', (byte)'t', (byte)'n', (byte)'u', (byte)'f', (byte)'r', (byte)'u', (byte)'u', // 00
                (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', (byte)'u', // 10
        };
    };

}   // namespace finalmq

File: src/json/JsonBuilder.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.


#include "finalmq/json/JsonBuilder.h"
#include "finalmq/conversions/itoa.h"
#include "finalmq/conversions/dtoa.h"
#include "finalmq/logger/LogStream.h"
#include "finalmq/helpers/ModulenameFinalmq.h"

#include <assert.h>
#include <string.h>
#include <iostream>


namespace finalmq {


JsonBuilder::JsonBuilder(IZeroCopyBuffer& buffer, int maxBlockSize)
    : m_zeroCopybuffer(buffer)
    , m_maxBlockSize(maxBlockSize)
{
}

JsonBuilder::~JsonBuilder()
{
    finished();
}


void JsonBuilder::reserveSpace(ssize_t space)
{
    ssize_t sizeRemaining = m_bufferEnd - m_buffer;
    if (sizeRemaining < space)
    {
        if (m_buffer != nullptr)
        {
            ssize_t size = m_buffer - m_bufferStart;
            assert(size >= 0);
            m_zeroCopybuffer.downsizeLastBuffer(size);
        }

        const ssize_t sizeRemainingZeroCopyBuffer = m_zeroCopybuffer.getRemainingSize();
        ssize_t sizeNew = m_maxBlockSize;
        if (space <= sizeRemainingZeroCopyBuffer)
        {
            sizeNew = sizeRemainingZeroCopyBuffer;
        }
        sizeNew = std::max(sizeNew, space);
        char* bufferStartNew = m_zeroCopybuffer.addBuffer(sizeNew);
        m_bufferStart = bufferStartNew;
        m_bufferEnd = m_bufferStart + sizeNew;
        m_buffer = m_bufferStart;
    }
}

void JsonBuilder::resizeBuffer()
{
    if (m_buffer != nullptr)
    {
        ssize_t size = m_buffer - m_bufferStart;
        assert(size >= 0);
        m_zeroCopybuffer.downsizeLastBuffer(size);
        m_bufferStart = nullptr;
        m_bufferEnd = nullptr;
        m_buffer = nullptr;
    }
}



// IJsonParserVisitor
void JsonBuilder::syntaxError(const char* /*str*/, const char* /*message*/)
{
    assert(false);
}

void JsonBuilder::enterNull()
{
    reserveSpace(5);
    assert(m_buffer);
    memcpy(m_buffer, "null,", 5);
    m_buffer += 5;
}

void JsonBuilder::enterBool(bool value)
{
    reserveSpace(6);
    assert(m_buffer);
    if (value)
    {
        memcpy(m_buffer, "true,", 5);
        m_buffer += 5;
    }
    else
    {
        memcpy(m_buffer, "false,", 6);
        m_buffer += 6;
    }
}

void JsonBuilder::enterInt32(std::int32_t value)
{
    reserveSpace(20);   // 11 + 1 is enough: 10 digits + 1 minus + 1 comma
    assert(m_buffer);
    m_buffer = rapidjson::i32toa(value, m_buffer);
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterUInt32(std::uint32_t value)
{
    reserveSpace(20);   // 10 + 1 is enough: 10 digits + 1 comma
    assert(m_buffer);
    m_buffer = rapidjson::u32toa(value, m_buffer);
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterInt64(std::int64_t value)
{
    reserveSpace(30);   // 21 + 1 is enough: 20 digits + 1 minus + 1 comma
    assert(m_buffer);
    m_buffer = rapidjson::i64toa(value, m_buffer);
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterUInt64(std::uint64_t value)
{
    reserveSpace(30);   // 20 + 1 is enough: 20 digits + 1 comma
    assert(m_buffer);
    m_buffer = rapidjson::u64toa(value, m_buffer);
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterDouble(double value)
{
    reserveSpace(50);
    assert(m_buffer);
    m_buffer = rapidjson::dtoa(value, m_buffer);
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterString(const char* value, ssize_t size)
{
    reserveSpace(size*6 + 2 + 1); // string*6 + 2" + comma
    *m_buffer = '\"';
    ++m_buffer;
    escapeString(value, size);
    *m_buffer = '\"';
    ++m_buffer;
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterString(std::string&& value)
{
    reserveSpace(value.size()*6 + 2 + 1); // string*6 + 2" + comma
    *m_buffer = '\"';
    ++m_buffer;
    escapeString(value.c_str(), value.size());
    *m_buffer = '\"';
    ++m_buffer;
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterArray()
{
    reserveSpace(1);
    *m_buffer = '[';
    ++m_buffer;
}

void JsonBuilder::exitArray()
{
    correctComma();
    reserveSpace(2);
    *m_buffer = ']';
    ++m_buffer;
    *m_buffer = ',';
    ++m_buffer;
}

void JsonBuilder::enterObject()
{
    reserveSpace(1);
    *m_buffer = '{';
    ++m_buffer;
}

void JsonBuilder::exitObject()
{
    correctComma();
    reserveSpace(2);
    *m_buffer = '}';
    ++m_buffer;
    *m_buffer = ',';
    ++m_buffer;
}




void JsonBuilder::enterKey(const char* key, ssize_t size)
{
    reserveSpace(size*6 + 2 + 1); // string*6 + 2" + :
    *m_buffer = '\"';
    ++m_buffer;
    escapeString(key, size);
    *m_buffer = '\"';
    ++m_buffer;
    *m_buffer = ':';
    ++m_buffer;
}

void JsonBuilder::enterKey(std::string&& key)
{
    reserveSpace(key.size()*6 + 2 + 1); // string*6 + 2" + :
    *m_buffer = '\"';
    ++m_buffer;
    escapeString(key.c_str(), key.size());
    *m_buffer = '\"';
    ++m_buffer;
    *m_buffer = ':';
    ++m_buffer;
}




void JsonBuilder::finished()
{
    correctComma();
    resizeBuffer();
}


void JsonBuilder::correctComma()
{
    if (m_buffer)
    {
        ssize_t size = m_buffer - m_bufferStart;
        assert(size >= 0);
        if (size >= 1)
        {
            if (*(m_buffer - 1) == ',')
            {
                --m_buffer;
            }
        }
    }
}



static unsigned char getRange(unsigned char c)
{
    // Referring to DFA of http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
    // With new mapping 1 -> 0x10, 7 -> 0x20, 9 -> 0x40, such that AND operation can test multiple types.
    static const unsigned char type[] =
    {
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x10,
        0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,0x40,
        0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,
        0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,
        8,8,2,2,2,2,2,2,2,2,2,2,2,2,2,2,  2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
        10,3,3,3,3,3,3,3,3,3,3,3,3,4,3,3, 11,6,6,6,5,8,8,8,8,8,8,8,8,8,8,8,
    };
    return type[c];
}




static bool decodeUtf8(const char*& str, unsigned int& codepoint)
{
#define RAPIDJSON_COPY() c = *str; str++; codepoint = (codepoint << 6) | (c & 0x3Fu)
#define RAPIDJSON_TRANS(mask) result &= ((getRange(c) & mask) != 0)
#define RAPIDJSON_TAIL() RAPIDJSON_COPY(); RAPIDJSON_TRANS(0x70)

    unsigned char c = static_cast<unsigned char>(*str);
    str++;
    if (!(c & 0x80))
    {
        codepoint = c;
        return true;
    }

    unsigned char type = getRange(c);
    if (type >= 32)
    {
        codepoint = 0;
    }
    else
    {
        codepoint = (0xFFu >> type) & c;
    }

    bool result = true;
    switch (type)
    {
    case 2: RAPIDJSON_TAIL(); return result;
    case 3: RAPIDJSON_TAIL(); RAPIDJSON_TAIL(); return result;
    case 4: RAPIDJSON_COPY(); RAPIDJSON_TRANS(0x50); RAPIDJSON_TAIL(); return result;
    case 5: RAPIDJSON_COPY(); RAPIDJSON_TRANS(0x10); RAPIDJSON_TAIL(); RAPIDJSON_TAIL(); return result;
    case 6: RAPIDJSON_TAIL(); RAPIDJSON_TAIL(); RAPIDJSON_TAIL(); return result;
    case 10: RAPIDJSON_COPY(); RAPIDJSON_TRANS(0x20); RAPIDJSON_TAIL(); return result;
    case 11: RAPIDJSON_COPY(); RAPIDJSON_TRANS(0x60); RAPIDJSON_TAIL(); RAPIDJSON_TAIL(); return result;
    default: return false;
    }
#undef RAPIDJSON_COPY
#undef RAPIDJSON_TRANS
#undef RAPIDJSON_TAIL
}






void JsonBuilder::escapeString(const char* str, ssize_t size)
{
    static const char hexDigits[16]{ '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };

    static const char escape[0x20] =
    {
            //0    1    2    3    4    5    6    7    8    9    A    B    C    D    E    F
            'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'b', 't', 'n', 'u', 'f', 'r', 'u', 'u', // 00
            'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', 'u', // 10
    };

    const char* end = str + size;


    while (str < end)
    {
        unsigned char c = static_cast<unsigned char>(*str);
        if (c < 0x20)
        {
            *m_buffer = '\\';
            m_buffer++;
            char e = escape[c];
            *m_buffer = e;
            m_buffer++;
            if (e == 'u')
            {
                *m_buffer = '0';
                m_buffer++;
                *m_buffer = '0';
                m_buffer++;
                *m_buffer = hexDigits[(c >> 4) & 0x0f];
                m_buffer++;
                *m_buffer = hexDigits[(c & 0x0f)];
                m_buffer++;
            }
            str++;
        }
        else if (c < 0x80)
        {
            if (c == '\\' || c == '\"')
            {
                *m_buffer = '\\';
                m_buffer++;
            }
            *m_buffer = c;
            m_buffer++;
            str++;
        }
        else
        {
            unsigned int codepoint;
            bool ok = decodeUtf8(str, codepoint);
            if (!ok)
            {
                streamError << "wrong utf8 format";
            }
            *m_buffer = '\\';
            m_buffer++;
            *m_buffer = 'u';
            m_buffer++;
            if (codepoint <= 0xD7FF || (codepoint >= 0xE000 && codepoint <= 0xFFFF)) {
                *m_buffer = hexDigits[((codepoint >> 12) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((codepoint >>  8) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((codepoint >>  4) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((codepoint)       & 0x0f)];
                m_buffer++;
            }
            else
            {
                assert(codepoint >= 0x010000 && codepoint <= 0x10FFFF);
                // Surrogate pair
                unsigned s = codepoint - 0x010000;
                unsigned lead = (s >> 10) + 0xD800;
                unsigned trail = (s & 0x3FF) + 0xDC00;
                *m_buffer = hexDigits[((lead >> 12) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((lead >>  8) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((lead >>  4) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((lead)       & 0x0f)];
                m_buffer++;
                *m_buffer = '\\';
                m_buffer++;
                *m_buffer = 'u';
                m_buffer++;
                *m_buffer = hexDigits[((trail >> 12) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((trail >>  8) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((trail >>  4) & 0x0f)];
                m_buffer++;
                *m_buffer = hexDigits[((trail)       & 0x0f)];
                m_buffer++;
            }
            str++;
        }
    }
}

}   // namespace finalmq
